velox/common/memory/tests/MappedMemoryTest.cpp

```diff
@@ -490,6 +490,40 @@ TEST_P(MappedMemoryTest, minSizeClass) {
   EXPECT_EQ(0, tracker->getCurrentUserBytes());
 }
 
+TEST_P(MappedMemoryTest, externalReserve) {
+  constexpr int32_t kReserveSizeUnit = 16;
+  constexpr int32_t kNumReserves = 100;
+  EXPECT_TRUE(instance_->checkConsistency());
+
+  uint64_t totalReserved = 0;
+  uint64_t counter = 0;
+  for (int i = 1; i <= kNumReserves; i++) {
+    auto toReserve = kReserveSizeUnit * i;
+    if (totalReserved + toReserve > kCapacity) {
+      if (useMmap_) {
+        EXPECT_FALSE(instance_->externalReserve(toReserve));
+        break;
+      }
+    }
+    EXPECT_TRUE(instance_->externalReserve(toReserve));
+    counter++;
+    totalReserved += toReserve;
+    EXPECT_EQ(instance_->numMapped(), totalReserved);
+    EXPECT_EQ(instance_->numAllocated(), totalReserved);
+    EXPECT_TRUE(instance_->checkConsistency());
+  }
+
+  for (int i = 1; i <= counter; i++) {
+    auto toReserve = kReserveSizeUnit * i;
+    EXPECT_NO_THROW(instance_->externalRelease(toReserve));
+    totalReserved -= toReserve;
+    EXPECT_EQ(instance_->numMapped(), totalReserved);
+    EXPECT_EQ(instance_->numAllocated(), totalReserved);
+    EXPECT_TRUE(instance_->checkConsistency());
```

Comment:
Could you also add to test through ScopedMappedMemory? Thanks!

Reply:
The entire suite tests 2 concrete implementations of MappedMemory -> MappedMemoryImpl and MmapAllocator. ScopedMappedMemory is a wrapper around either of these and delegates essentially all calls to MappedMemoryImpl and MmapAllocator. There is a test that specifically tests ScopedMappedMemory, I think I can add a couple of reservation calls there. That might be a better place.